x-pack/plugin/repository-encrypted/src/test/java/org/elasticsearch/repositories/encrypted/ChainingInputStreamTests.java

```diff
@@ -0,0 +1,573 @@
+        test.available();
+        verify(mockCurrentIn).available();
+    }
+
+    public void testEmptyReadAsksForNext() throws Exception {
+        InputStream mockCurrentIn = mock(InputStream.class);
+        when(mockCurrentIn.markSupported()).thenAnswer(invocationOnMock -> true);
+        ChainingInputStream test = new ChainingInputStream() {
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                return mockCurrentIn;
+            }
+        };
+        test.currentIn = InputStream.nullInputStream();
+        when(mockCurrentIn.read()).thenAnswer(invocationOnMock -> Randomness.get().nextInt(256));
+        test.read();
+        verify(mockCurrentIn).read();
+        // test "array read"
+        test.currentIn = InputStream.nullInputStream();
+        when(mockCurrentIn.read(org.mockito.Matchers.<byte[]>any(), org.mockito.Matchers.anyInt(), org.mockito.Matchers.anyInt())).
+                thenAnswer(invocationOnMock -> {
+                    final int len = (int) invocationOnMock.getArguments()[2];
+                    if (len == 0) {
+                        return 0;
+                    } else {
+                        int bytesCount = 1 + Randomness.get().nextInt(len);
+                        return bytesCount;
+                    }
+                });
+        byte[] b = new byte[2 + Randomness.get().nextInt(32)];
+        int len = 1 + Randomness.get().nextInt(b.length - 1);
+        int offset = Randomness.get().nextInt(b.length - len);
+        test.read(b, offset, len);
+        verify(mockCurrentIn).read(Mockito.eq(b), Mockito.eq(offset), Mockito.eq(len));
+    }
+
+    public void testReadAll() throws Exception {
+        byte[] b = new byte[2 + Randomness.get().nextInt(32)];
+        Randomness.get().nextBytes(b);
+        int splitIdx = Randomness.get().nextInt(b.length - 1);
+        ByteArrayInputStream first = new ByteArrayInputStream(b, 0, splitIdx + 1);
+        ByteArrayInputStream second = new ByteArrayInputStream(b, splitIdx + 1, b.length - splitIdx - 1);
+        ChainingInputStream test = new ChainingInputStream() {
+            @Override
+            InputStream nextComponent(InputStream currentElementIn) throws IOException {
+                if (currentElementIn == null) {
+                    return first;
+                } else if (currentElementIn == first) {
+                    return second;
+                } else if (currentElementIn == second) {
+                    return null;
+                } else {
+                    throw new IllegalArgumentException();
+                }
+            }
+        };
+        byte[] result = test.readAllBytes();
+        assertThat(result.length, Matchers.is(b.length));
+        for (int i = 0; i < result.length; i++) {
+            Assert.assertThat(result[i], Matchers.is(b[i]));
+        }
+    }
+
+    public void testMark() throws Exception {
```

Comment:
Thank you for taking the time to think though the testing of it! I really appreciate it.

I agree in principle, and I will work to address the cases as you outlined.

But, I want to raise a couple of things that will downplay the supposed benefits of the outline you put up:

1) the `testMark` and `testReset` cases are much more powerful than they look at first sight. All the `mark` cases that you illustrated work the same way: a reference to the current component is stored internally and the mark call is forwarded to that component. There is no outward result to test/assert. The overall position among components and inside a component is irrelevant. This is analogous for `reset`: no matter how many bytes and components have been read/skipped since the last mark, the `reset` call makes the stored reference the current component, and then forwards the reset call to that one component. The previous current component is simply discarded (possibly closed).
These two test methods, `testMark` and `testReset`, are testing the theoretical correctness of the internal state. They cover a lot of ground, but they are brittle if the implementation of the `ChainingInputStream` changes.
On a more metaphysical level, for better or worst, this is the aimed result of creating the `ChainingInputStream` component in the first place, rather than coding the `EncryptionInputStream` in one piece.

2) The practical testing of the actual output in a `mark` / `reset` scenario is hard to debug very soon and any actual bug won't show up in a single test failure, most of the ones in the outline will fail for the simplest mistake (i.e. they overlap a great deal, the internal state is not different among those cases).

3) This does not consider the `mark` after a `reset` (as different from `mark` after `mark`) which I think is a very treacherous flow. And there are very many cases in that flow, e.g. the mark after a reset is before/beyond the first mark `x` the first mark is on the current/border/after the current component.